File: api/Enkl.Api/Services/StrategyPillarService.cs
using Enkl.Api.Data;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

/// <summary>
/// Pillar + Enabler CRUD combined in one service — Enabler is a trivial child of Pillar, split
/// further only if this file gets large (same don't-over-split judgment as PortfolioCategoryService/
/// PortfolioResourceService's own split, root CLAUDE.md's api-tier remediation notes). Every method
/// re-validates ownership up the chain (Pillar -> Strategy -> Organisation, Enabler -> Pillar ->
/// Strategy -> Organisation) before touching anything — same "second FK gets its own re-validation"
/// discipline as PortfolioService.UpdateProjectCategoryAsync.
/// </summary>
public class StrategyPillarService
{
    private readonly AppDbContext _db;

    public StrategyPillarService(AppDbContext db)
    {
        _db = db;
    }

    private Task<bool> StrategyBelongsToOrgAsync(Guid organisationId, Guid strategyId) =>
        _db.Strategies.AnyAsync(s => s.Id == strategyId && s.OrganisationId == organisationId);

    private async Task<StrategyPillar?> GetOwnedPillarAsync(Guid organisationId, Guid pillarId)
    {
        return await _db.StrategyPillars
            .Include(p => p.Strategy)
            .FirstOrDefaultAsync(p => p.Id == pillarId && p.Strategy.OrganisationId == organisationId);
    }

    private async Task<StrategyEnabler?> GetOwnedEnablerAsync(Guid organisationId, Guid enablerId)
    {
        return await _db.StrategyEnablers
            .Include(e => e.Pillar).ThenInclude(p => p.Strategy)
            .FirstOrDefaultAsync(e => e.Id == enablerId && e.Pillar.Strategy.OrganisationId == organisationId);
    }

    // ---- Pillars ----

    public async Task<StrategyPillarDto?> CreatePillarAsync(Guid organisationId, Guid strategyId, CreateStrategyPillarRequest request)
    {
        if (!await StrategyBelongsToOrgAsync(organisationId, strategyId)) return null;

        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) return null;
        if (name.Length > 150) name = name[..150];

        var maxSortOrder = await _db.StrategyPillars.Where(p => p.StrategyId == strategyId)
            .Select(p => (int?)p.SortOrder).MaxAsync() ?? -1;

        var pillar = new StrategyPillar
        {
            Id = Guid.NewGuid(),
            StrategyId = strategyId,
            Name = name,
            Description = string.IsNullOrWhiteSpace(request.Description) ? null : request.Description!.Trim(),
            SortOrder = maxSortOrder + 1
        };
        _db.StrategyPillars.Add(pillar);
        await _db.SaveChangesAsync();
        return ToDto(pillar);
    }

    public async Task<StrategyPillarDto?> UpdatePillarAsync(Guid organisationId, Guid pillarId, UpdateStrategyPillarRequest request)
    {
        var pillar = await GetOwnedPillarAsync(organisationId, pillarId);
        if (pillar is null) return null;

        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) return null;
        if (name.Length > 150) name = name[..150];

        pillar.Name = name;
        pillar.Description = string.IsNullOrWhiteSpace(request.Description) ? null : request.Description!.Trim();
        pillar.SortOrder = request.SortOrder;
        await _db.SaveChangesAsync();
        return ToDto(pillar);
    }

    public async Task<bool> DeletePillarAsync(Guid organisationId, Guid pillarId)
    {
        var pillar = await GetOwnedPillarAsync(organisationId, pillarId);
        if (pillar is null) return false;

        _db.StrategyPillars.Remove(pillar);
        await _db.SaveChangesAsync();
        return true;
    }

    // ---- Enablers ----

    public async Task<StrategyEnablerDto?> CreateEnablerAsync(Guid organisationId, Guid pillarId, CreateStrategyEnablerRequest request)
    {
        var pillar = await GetOwnedPillarAsync(organisationId, pillarId);
        if (pillar is null) return null;

        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) return null;
        if (name.Length > 150) name = name[..150];

        var maxSortOrder = await _db.StrategyEnablers.Where(e => e.PillarId == pillarId)
            .Select(e => (int?)e.SortOrder).MaxAsync() ?? -1;

        var enabler = new StrategyEnabler
        {
            Id = Guid.NewGuid(),
            PillarId = pillarId,
            Name = name,
            Description = string.IsNullOrWhiteSpace(request.Description) ? null : request.Description!.Trim(),
            SortOrder = maxSortOrder + 1
        };
        _db.StrategyEnablers.Add(enabler);
        await _db.SaveChangesAsync();
        return ToDto(enabler);
    }

    public async Task<StrategyEnablerDto?> UpdateEnablerAsync(Guid organisationId, Guid enablerId, UpdateStrategyEnablerRequest request)
    {
        var enabler = await GetOwnedEnablerAsync(organisationId, enablerId);
        if (enabler is null) return null;

        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) return null;
        if (name.Length > 150) name = name[..150];

        enabler.Name = name;
        enabler.Description = string.IsNullOrWhiteSpace(request.Description) ? null : request.Description!.Trim();
        enabler.SortOrder = request.SortOrder;
        await _db.SaveChangesAsync();
        return ToDto(enabler);
    }

    public async Task<bool> DeleteEnablerAsync(Guid organisationId, Guid enablerId)
    {
        var enabler = await GetOwnedEnablerAsync(organisationId, enablerId);
        if (enabler is null) return false;

        _db.StrategyEnablers.Remove(enabler);
        await _db.SaveChangesAsync();
        return true;
    }

    // ---- Tree read (Pillars -> Enablers -> Metrics, plus Metrics directly on a Pillar) ----
    // Shared by StrategyController (OrgAdmin) and ProjectStrategyController (ProjectMember, read-only)
    // — same shape both callers need, queried directly here rather than round-tripping through
    // StrategyMetricService, same "a service freely queries another entity's table when it needs to
    // assemble one composite read" convention as PortfolioResourceService.GetResourcingSummaryAsync.

    public async Task<List<StrategyPillarTreeDto>> GetPillarTreeAsync(Guid strategyId)
    {
        var pillars = await _db.StrategyPillars.AsNoTracking()
            .Where(p => p.StrategyId == strategyId)
            .OrderBy(p => p.SortOrder)
            .ToListAsync();
        var pillarIds = pillars.Select(p => p.Id).ToList();

        var enablers = await _db.StrategyEnablers.AsNoTracking()
            .Where(e => pillarIds.Contains(e.PillarId))
            .OrderBy(e => e.SortOrder)
            .ToListAsync();
        var enablerIds = enablers.Select(e => e.Id).ToList();

        var metrics = await _db.StrategyMetrics.AsNoTracking()
            .Where(m => (m.PillarId != null && pillarIds.Contains(m.PillarId.Value)) || (m.EnablerId != null && enablerIds.Contains(m.EnablerId.Value)))
            .OrderBy(m => m.SortOrder)
            .ToListAsync();

        List<StrategyMetricDto> MetricsFor(Func<StrategyMetric, bool> predicate) =>
            metrics.Where(predicate).Select(m => new StrategyMetricDto(m.Id, m.PillarId, m.EnablerId, m.Name, m.TargetValue, m.UnitLabel, m.SortOrder)).ToList();

        return pillars.Select(p => new StrategyPillarTreeDto(
            p.Id, p.Name, p.Description, p.SortOrder,
            MetricsFor(m => m.PillarId == p.Id),
            enablers.Where(e => e.PillarId == p.Id).Select(e => new StrategyEnablerTreeDto(
                e.Id, e.Name, e.Description, e.SortOrder,
                MetricsFor(m => m.EnablerId == e.Id)
            )).ToList()
        )).ToList();
    }

    /// <summary>ProjectMember-readable variant of GetPillarTreeAsync — resolves the project's own
    /// org, then that org's active Strategy, then the tree; used by ProjectStrategyController where
    /// the caller has no strategyId of their own to supply. Null means either the project doesn't
    /// exist or the org has no active Strategy yet (both collapse to the same "nothing to show"
    /// response, no enumeration oracle needed here since this is a read-only, non-sensitive surface).</summary>
    public async Task<List<StrategyPillarTreeDto>?> GetActivePillarTreeForProjectAsync(Guid projectId)
    {
        var organisationId = await _db.Projects.Where(p => p.Id == projectId).Select(p => (Guid?)p.OrganisationId).FirstOrDefaultAsync();
        if (organisationId is null) return null;

        var activeStrategyId = await _db.Strategies
            .Where(s => s.OrganisationId == organisationId && s.IsActive)
            .Select(s => (Guid?)s.Id).FirstOrDefaultAsync();
        if (activeStrategyId is null) return null;

        return await GetPillarTreeAsync(activeStrategyId.Value);
    }

    private static StrategyPillarDto ToDto(StrategyPillar p) => new(p.Id, p.StrategyId, p.Name, p.Description, p.SortOrder);
    private static StrategyEnablerDto ToDto(StrategyEnabler e) => new(e.Id, e.PillarId, e.Name, e.Description, e.SortOrder);
}

File: php-api/src/Services/StrategyPillarService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Support\Uuid;
use PDO;

/**
 * Ported from Services/StrategyPillarService.cs. Pillar + Enabler CRUD combined in one class — same
 * don't-over-split judgment as the .NET tier. Every method re-validates ownership up the chain
 * (Pillar -> Strategy -> Organisation, Enabler -> Pillar -> Strategy -> Organisation) before touching
 * anything.
 */
final class StrategyPillarService
{
    public function __construct(private readonly PDO $db)
    {
    }

    private function strategyBelongsToOrg(string $organisationId, string $strategyId): bool
    {
        $stmt = $this->db->prepare('SELECT 1 FROM "Strategies" WHERE "Id" = :id AND "OrganisationId" = :orgId');
        $stmt->execute(['id' => $strategyId, 'orgId' => $organisationId]);
        return $stmt->fetch() !== false;
    }

    private function findOwnedPillar(string $organisationId, string $pillarId): ?array
    {
        $stmt = $this->db->prepare(<<<SQL
            SELECT p.* FROM "StrategyPillars" p
            JOIN "Strategies" s ON s."Id" = p."StrategyId"
            WHERE p."Id" = :id AND s."OrganisationId" = :orgId
        SQL);
        $stmt->execute(['id' => $pillarId, 'orgId' => $organisationId]);
        $row = $stmt->fetch();
        return $row !== false ? $row : null;
    }

    private function findOwnedEnabler(string $organisationId, string $enablerId): ?array
    {
        $stmt = $this->db->prepare(<<<SQL
            SELECT e.* FROM "StrategyEnablers" e
            JOIN "StrategyPillars" p ON p."Id" = e."PillarId"
            JOIN "Strategies" s ON s."Id" = p."StrategyId"
            WHERE e."Id" = :id AND s."OrganisationId" = :orgId
        SQL);
        $stmt->execute(['id' => $enablerId, 'orgId' => $organisationId]);
        $row = $stmt->fetch();
        return $row !== false ? $row : null;
    }

    // ---- Pillars ----

    public function createPillar(string $organisationId, string $strategyId, array $request): ?array
    {
        if (!$this->strategyBelongsToOrg($organisationId, $strategyId)) {
            return null;
        }

        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            return null;
        }
        if (strlen($name) > 150) {
            $name = substr($name, 0, 150);
        }
        $description = trim((string) ($request['description'] ?? '')) !== '' ? trim((string) $request['description']) : null;

        $maxStmt = $this->db->prepare('SELECT MAX("SortOrder") FROM "StrategyPillars" WHERE "StrategyId" = :sid');
        $maxStmt->execute(['sid' => $strategyId]);
        $max = $maxStmt->fetchColumn();
        $sortOrder = ($max !== false && $max !== null) ? ((int) $max) + 1 : 0;

        $pillarId = Uuid::v4();
        $this->db->prepare('INSERT INTO "StrategyPillars" ("Id", "StrategyId", "Name", "Description", "SortOrder") VALUES (:id, :sid, :name, :description, :sortOrder)')
            ->execute(['id' => $pillarId, 'sid' => $strategyId, 'name' => $name, 'description' => $description, 'sortOrder' => $sortOrder]);

        return ['id' => $pillarId, 'strategyId' => $strategyId, 'name' => $name, 'description' => $description, 'sortOrder' => $sortOrder];
    }

    public function updatePillar(string $organisationId, string $pillarId, array $request): ?array
    {
        $pillar = $this->findOwnedPillar($organisationId, $pillarId);
        if ($pillar === null) {
            return null;
        }

        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            return null;
        }
        if (strlen($name) > 150) {
            $name = substr($name, 0, 150);
        }
        $description = trim((string) ($request['description'] ?? '')) !== '' ? trim((string) $request['description']) : null;
        $sortOrder = (int) ($request['sortOrder'] ?? $pillar['SortOrder']);

        $this->db->prepare('UPDATE "StrategyPillars" SET "Name" = :name, "Description" = :description, "SortOrder" = :sortOrder WHERE "Id" = :id')
            ->execute(['name' => $name, 'description' => $description, 'sortOrder' => $sortOrder, 'id' => $pillarId]);

        return ['id' => $pillarId, 'strategyId' => $pillar['StrategyId'], 'name' => $name, 'description' => $description, 'sortOrder' => $sortOrder];
    }

    public function deletePillar(string $organisationId, string $pillarId): bool
    {
        if ($this->findOwnedPillar($organisationId, $pillarId) === null) {
            return false;
        }
        $this->db->prepare('DELETE FROM "StrategyPillars" WHERE "Id" = :id')->execute(['id' => $pillarId]);
        return true;
    }

    // ---- Enablers ----

    public function createEnabler(string $organisationId, string $pillarId, array $request): ?array
    {
        if ($this->findOwnedPillar($organisationId, $pillarId) === null) {
            return null;
        }

        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            return null;
        }
        if (strlen($name) > 150) {
            $name = substr($name, 0, 150);
        }
        $description = trim((string) ($request['description'] ?? '')) !== '' ? trim((string) $request['description']) : null;

        $maxStmt = $this->db->prepare('SELECT MAX("SortOrder") FROM "StrategyEnablers" WHERE "PillarId" = :pid');
        $maxStmt->execute(['pid' => $pillarId]);
        $max = $maxStmt->fetchColumn();
        $sortOrder = ($max !== false && $max !== null) ? ((int) $max) + 1 : 0;

        $enablerId = Uuid::v4();
        $this->db->prepare('INSERT INTO "StrategyEnablers" ("Id", "PillarId", "Name", "Description", "SortOrder") VALUES (:id, :pid, :name, :description, :sortOrder)')
            ->execute(['id' => $enablerId, 'pid' => $pillarId, 'name' => $name, 'description' => $description, 'sortOrder' => $sortOrder]);

        return ['id' => $enablerId, 'pillarId' => $pillarId, 'name' => $name, 'description' => $description, 'sortOrder' => $sortOrder];
    }

    public function updateEnabler(string $organisationId, string $enablerId, array $request): ?array
    {
        $enabler = $this->findOwnedEnabler($organisationId, $enablerId);
        if ($enabler === null) {
            return null;
        }

        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            return null;
        }
        if (strlen($name) > 150) {
            $name = substr($name, 0, 150);
        }
        $description = trim((string) ($request['description'] ?? '')) !== '' ? trim((string) $request['description']) : null;
        $sortOrder = (int) ($request['sortOrder'] ?? $enabler['SortOrder']);

        $this->db->prepare('UPDATE "StrategyEnablers" SET "Name" = :name, "Description" = :description, "SortOrder" = :sortOrder WHERE "Id" = :id')
            ->execute(['name' => $name, 'description' => $description, 'sortOrder' => $sortOrder, 'id' => $enablerId]);

        return ['id' => $enablerId, 'pillarId' => $enabler['PillarId'], 'name' => $name, 'description' => $description, 'sortOrder' => $sortOrder];
    }

    public function deleteEnabler(string $organisationId, string $enablerId): bool
    {
        if ($this->findOwnedEnabler($organisationId, $enablerId) === null) {
            return false;
        }
        $this->db->prepare('DELETE FROM "StrategyEnablers" WHERE "Id" = :id')->execute(['id' => $enablerId]);
        return true;
    }

    // ---- Tree read (Pillars -> Enablers -> Metrics, plus Metrics directly on a Pillar) ----
    // Shared by StrategyController (OrgAdmin) and ProjectStrategyController (ProjectMember, read-only).

    public function getPillarTree(string $strategyId): array
    {
        $pillarStmt = $this->db->prepare('SELECT * FROM "StrategyPillars" WHERE "StrategyId" = :sid ORDER BY "SortOrder"');
        $pillarStmt->execute(['sid' => $strategyId]);
        $pillars = $pillarStmt->fetchAll();
        $pillarIds = array_column($pillars, 'Id');

        $enablers = [];
        if (count($pillarIds) > 0) {
            $placeholders = implode(',', array_map(static fn(int $i): string => ":pid{$i}", array_keys($pillarIds)));
            $stmt = $this->db->prepare("SELECT * FROM \"StrategyEnablers\" WHERE \"PillarId\" IN ({$placeholders}) ORDER BY \"SortOrder\"");
            $params = [];
            foreach ($pillarIds as $i => $id) {
                $params["pid{$i}"] = $id;
            }
            $stmt->execute($params);
            $enablers = $stmt->fetchAll();
        }
        $enablerIds = array_column($enablers, 'Id');

        $metrics = [];
        if (count($pillarIds) > 0 || count($enablerIds) > 0) {
            $conditions = [];
            $params = [];
            if (count($pillarIds) > 0) {
                $ph = implode(',', array_map(static fn(int $i): string => ":pid{$i}", array_keys($pillarIds)));
                $conditions[] = "\"PillarId\" IN ({$ph})";
                foreach ($pillarIds as $i => $id) {
                    $params["pid{$i}"] = $id;
                }
            }
            if (count($enablerIds) > 0) {
                $ph = implode(',', array_map(static fn(int $i): string => ":eid{$i}", array_keys($enablerIds)));
                $conditions[] = "\"EnablerId\" IN ({$ph})";
                foreach ($enablerIds as $i => $id) {
                    $params["eid{$i}"] = $id;
                }
            }
            $stmt = $this->db->prepare('SELECT * FROM "StrategyMetrics" WHERE ' . implode(' OR ', $conditions) . ' ORDER BY "SortOrder"');
            $stmt->execute($params);
            $metrics = $stmt->fetchAll();
        }

        $metricDto = static fn(array $m): array => [
            'id' => $m['Id'], 'pillarId' => $m['PillarId'], 'enablerId' => $m['EnablerId'], 'name' => $m['Name'],
            'targetValue' => $m['TargetValue'] !== null ? (float) $m['TargetValue'] : null,
            'unitLabel' => $m['UnitLabel'], 'sortOrder' => (int) $m['SortOrder'],
        ];
        $metricsFor = static fn(string $key, string $value): array =>
            array_values(array_map($metricDto, array_filter($metrics, static fn(array $m) => $m[$key] === $value)));

        return array_map(static function (array $p) use ($enablers, $metricsFor, $metricDto): array {
            $pillarEnablers = array_values(array_filter($enablers, static fn(array $e) => $e['PillarId'] === $p['Id']));
            return [
                'id' => $p['Id'], 'name' => $p['Name'], 'description' => $p['Description'], 'sortOrder' => (int) $p['SortOrder'],
                'metrics' => $metricsFor('PillarId', $p['Id']),
                'enablers' => array_map(static fn(array $e) => [
                    'id' => $e['Id'], 'name' => $e['Name'], 'description' => $e['Description'], 'sortOrder' => (int) $e['SortOrder'],
                    'metrics' => $metricsFor('EnablerId', $e['Id']),
                ], $pillarEnablers),
            ];
        }, $pillars);
    }

    /** ProjectMember-readable variant — resolves the project's own org, then that org's active
     * Strategy, then the tree; null means either the project doesn't exist or the org has no active
     * Strategy yet (both collapse to "nothing to show", no enumeration oracle needed on this
     * read-only, non-sensitive surface). */
    public function getActivePillarTreeForProject(string $projectId): ?array
    {
        $stmt = $this->db->prepare('SELECT "OrganisationId" FROM "Projects" WHERE "Id" = :id');
        $stmt->execute(['id' => $projectId]);
        $organisationId = $stmt->fetchColumn();
        if ($organisationId === false) {
            return null;
        }

        $stmt = $this->db->prepare('SELECT "Id" FROM "Strategies" WHERE "OrganisationId" = :orgId AND "IsActive" = true');
        $stmt->execute(['orgId' => $organisationId]);
        $strategyId = $stmt->fetchColumn();
        if ($strategyId === false) {
            return null;
        }

        return $this->getPillarTree($strategyId);
    }
}
